coev/evidence_id.go

```diff
@@ -0,0 +1,199 @@
+	if !ok {
+		return nil, fmt.Errorf("unknown EvidenceID type: %s", typ)
+	}
+
+	return factory(val)
+}
+
+// Valid checks for the validity of given EvidenceID
+func (o EvidenceID) Valid() error {
+	if o.String() == "" {
+		return errors.New("no EvidenceID")
+	}
+	_, err := o.GetUUID()
+	if err != nil {
+		return fmt.Errorf("unable to fetch valid UUID: %w", err)
+	}
+	return nil
+}
+
+// String returns a printable string of the EvidenceID value.  UUIDs use the
+// canonical 8-4-4-4-12 format.
+func (o EvidenceID) String() string {
+	if o.Value == nil {
+		return ""
+	}
+
+	return o.Value.String()
+}
+
+// Type returns a string naming the type of the underlying EvidenceID value.
+func (o EvidenceID) Type() string {
+	return o.Value.Type()
+}
+
+// Bytes returns a []byte containing the bytes of the underlying EvidenceID
+// value.
+func (o EvidenceID) Bytes() []byte {
+	return o.Value.Bytes()
+}
+
+// MarshalCBOR serializes the target EvidenceID to CBOR
+func (o EvidenceID) MarshalCBOR() ([]byte, error) {
+	return em.Marshal(o.Value)
+}
+
+func (o *EvidenceID) UnmarshalCBOR(data []byte) error {
+	return dm.Unmarshal(data, &o.Value)
+}
+
+// UnmarshalJSON deserializes the supplied JSON object into the target EvidenceID
+// The evidence object must have the following shape:
+//
+//	{
+//	  "type": "<EVIDENCE_TYPE>",
+//	  "value": <EVIDENCE_VALUE>
+//	}
+//
+// where <EVIDENCE_TYPE> must be one of the known IEvidenceValue implementation
+// type names (available in the base implementation: "uuid"), and
+// <EVIDENCE_VALUE> is the JSON encoding of the evidence value. The exact
+// encoding is <EVIDENCE_TYPE> dependent. For the base implmentation types it is
+//
+//	uuid: standard UUID string representation, e.g. "550e8400-e29b-41d4-a716-446655440000"
+
+//nolint:dupl
+func (o *EvidenceID) UnmarshalJSON(data []byte) error {
+	var tnv encoding.TypeAndValue
+
+	if err := json.Unmarshal(data, &tnv); err != nil {
+		return fmt.Errorf("EvidenceID decoding failure: %w", err)
+	}
+
+	decoded, err := NewEvidenceID(nil, tnv.Type)
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(tnv.Value, &decoded.Value); err != nil {
+		return fmt.Errorf(
+			"cannot unmarshal EvidenceID: %w",
+			err,
+		)
+	}
+
+	if err := decoded.Value.Valid(); err != nil {
+		return fmt.Errorf("invalid %s: %w", tnv.Type, err)
+	}
+
+	o.Value = decoded.Value
+
+	return nil
+}
+
+// MarshalJSON serializes the EvidenceID into a JSON object.
+func (o EvidenceID) MarshalJSON() ([]byte, error) {
+	valueBytes, err := json.Marshal(o.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	value := encoding.TypeAndValue{
+		Type:  o.Value.Type(),
+		Value: valueBytes,
+	}
+
+	return json.Marshal(value)
+}
+
+// SetUUID sets the identity of the target Evidence to the supplied UUID
+func (o *EvidenceID) SetUUID(val uuid.UUID) *EvidenceID {
+	if o != nil {
+		o.Value = comid.TaggedUUID(val)
+	}
+	return o
+}
+
+func (o EvidenceID) GetUUID() (comid.UUID, error) {
```

Comment:
missing documentation of a public function

Reply:
Done added new comment to clarify the usage!